### DISCUSSION TITLE
How to make systems really  multithreaded?

### DISCUSSION THREAD

**Discussion by robo-todd** [SanderMertens/flecs#1136] [Q&A]
First, I want to say that flecs is amazing. I'm using it for non-game applications and I think it is a revolution in how I can build code. I am trying (learning) how to make my system highly biased towards being entity parallel, and I'm starting to use multi threaded systems.

 I have turned on multiple threads  (4 in my test case) at the top of the loop, and I have some multi-threaded systems that have very heavy processing in their eval() functions. I had hoped, that each such multi-threaded system would run on a worker thread, but even though they are marked multi_threaded() they just run sequentially on the same thread (printing out the thread id in the eval call...).
There are only a small number of entities (approx the number of threads) that match these systems, and the system query does use/follow relationships to pull in inputs (read), but the output data are updating separate entities in eval.

Any pointers on how to really make the systems run on multiple threads? I've read through the docs quite a few times and I *think* I'm doing it right, but no luck. They don't seem to get scheduled to run on separate threads.

This would be tremendously powerful if I can get it working or understand what would make a multi_threaded() system *not* run multi threaded. Perhaps another way to ask the question is what would cause a multi-threaded system to be run single threaded?

I'm using C++ and I'm building from flecs v3.2.10 release code.

**Comment by copygirl** [comment]
Just to make sure, you also called `world.set_threads()`, right? (As mentioned in the docs [here](https://www.flecs.dev/flecs/md_docs_2Systems.html#threading).) I believe systems will only be run simultaneously if they don't access the same components with write access, that means marking them as `[in]` in queries.

**Reply by roboticsgeek** [reply]
Yes. Called world.set_threads(4) before the loop.  
I'm not sure I understand the second part. The system writes a component, let's call it "foo".. that is part of the match query and is marked out. The idea is that each separate entity would be processed on a separate thread and each write their separate foo (same component) at the same time... at least that's how I understood the documentation on multi-threading. This is a single system, marked multi-threaded, and my intention was to have it run entity-parallel (If I'm understanding this all correctly.)

**Reply by robo-todd** [reply]
Additional information that might be relevant: The system also matches a singleton "in". Could that be tripping it up?

**Reply by SanderMertens** [reply]
Are you using `app().run()` by any chance? Found an issue recently where that prevents running on multiple threads (still need to fix that).

**Reply by robo-todd** [reply]
I'm not using app.run, just the "progress loop" style setup for the world. But I still have this issue and I've changed a lot of things and I can't figure out why it won't run in parallel. Due to workload it will take some time for me, but I want to come back and try to post a simplified example that shows what I mean.

**Reply by robo-todd** [reply]
I finally got enough time to make an example of what I'm talking about. Attached is a C++ source file that I'm building against flecs 3.2.10. If you provide no command-line argument it runs single threaded. If you provide a command line argument it turns on 4 threads in an attempt to be entity parallel. However, I am somewhat confused by how it does this. It does seem to run this example on separate threads, but I would have expected all the threads to complete a phase and then all the threads to work on the next phase, etc. When threaded it seems entity parallel, but it runs phases independently for each entity in parallel.  So, I'm trying to understand fully how multi-threading/entity parallelism works.   If an entity has to have a clean trace of in-out across all phases, I don't see how I can get much parallelism in a larger system, but if entities are parallel within a phase then that works well (for what I'm trying to do at least). Am I misunderstanding something here?

Non threading output:
===================== START ================
    Main Thread ID: 140415341217600
UpdateData for entity: Test.0 on thread 140415341217600
UpdateData for entity: Test.1 on thread 140415341217600
UpdateData for entity: Test.2 on thread 140415341217600
UpdateData for entity: Test.3 on thread 140415341217600
UpdateData for entity: Test.4 on thread 140415341217600
UpdateData for entity: Test.5 on thread 140415341217600
UpdateData for entity: Test.6 on thread 140415341217600
UpdateData for entity: Test.7 on thread 140415341217600
ComputeResult for entity: Test.0 on thread 140415341217600
ComputeResult for entity: Test.1 on thread 140415341217600
ComputeResult for entity: Test.2 on thread 140415341217600
ComputeResult for entity: Test.3 on thread 140415341217600
ComputeResult for entity: Test.4 on thread 140415341217600
ComputeResult for entity: Test.5 on thread 140415341217600
ComputeResult for entity: Test.6 on thread 140415341217600
ComputeResult for entity: Test.7 on thread 140415341217600
@ Iteration: 1
UpdateData for entity: Test.0 on thread 140415341217600
UpdateData for entity: Test.1 on thread 140415341217600
UpdateData for entity: Test.2 on thread 140415341217600
UpdateData for entity: Test.3 on thread 140415341217600
UpdateData for entity: Test.4 on thread 140415341217600
UpdateData for entity: Test.5 on thread 140415341217600
UpdateData for entity: Test.6 on thread 140415341217600
UpdateData for entity: Test.7 on thread 140415341217600
ComputeResult for entity: Test.0 on thread 140415341217600
ComputeResult for entity: Test.1 on thread 140415341217600
ComputeResult for entity: Test.2 on thread 140415341217600
ComputeResult for entity: Test.3 on thread 140415341217600
ComputeResult for entity: Test.4 on thread 140415341217600
ComputeResult for entity: Test.5 on thread 140415341217600
ComputeResult for entity: Test.6 on thread 140415341217600
ComputeResult for entity: Test.7 on thread 140415341217600
@ Iteration: 2
UpdateData for entity: Test.0 on thread 140415341217600
UpdateData for entity: Test.1 on thread 140415341217600
UpdateData for entity: Test.2 on thread 140415341217600
UpdateData for entity: Test.3 on thread 140415341217600
UpdateData for entity: Test.4 on thread 140415341217600
UpdateData for entity: Test.5 on thread 140415341217600
UpdateData for entity: Test.6 on thread 140415341217600
UpdateData for entity: Test.7 on thread 140415341217600
ComputeResult for entity: Test.0 on thread 140415341217600
ComputeResult for entity: Test.1 on thread 140415341217600
ComputeResult for entity: Test.2 on thread 140415341217600
ComputeResult for entity: Test.3 on thread 140415341217600
ComputeResult for entity: Test.4 on thread 140415341217600
ComputeResult for entity: Test.5 on thread 140415341217600
ComputeResult for entity: Test.6 on thread 140415341217600
ComputeResult for entity: Test.7 on thread 140415341217600
=============== Thread Test Exit ============


Multi-Threading Output:
===================== START ================
    Main Thread ID: 139882963007296
UpdateData for entity: Test.0 on thread 139882963007296
UpdateData for entity: Test.1 on thread 139882963007296
ComputeResult for entity: Test.0 on thread 139882963007296
ComputeResult for entity: Test.1 on thread 139882963007296
UpdateData for entity: Test.6 on thread 139882938156608
UpdateData for entity: Test.7 on thread 139882938156608
ComputeResult for entity: Test.6 on thread 139882938156608
ComputeResult for entity: Test.7 on thread 139882938156608
UpdateData for entity: UpdateData for entity: Test.2 on thread 139882954942016
UpdateData for entity: Test.3 on thread 139882954942016
ComputeResult for entity: Test.2 on thread 139882954942016
Test.4 on thread 139882946549312
UpdateData for entity: Test.5 on thread 139882946549312
ComputeResult for entity: Test.3 on thread 139882954942016
ComputeResult for entity: Test.4 on thread 139882946549312
ComputeResult for entity: Test.5 on thread 139882946549312
@ Iteration: 1
UpdateData for entity: Test.0 on thread 139882963007296
UpdateData for entity: Test.1 on thread 139882963007296
ComputeResult for entity: Test.0 on thread 139882963007296
ComputeResult for entity: Test.1 on thread 139882963007296
UpdateData for entity: Test.6 on thread 139882938156608
UpdateData for entity: Test.7 on thread 139882938156608
ComputeResult for entity: Test.6 on thread 139882938156608
ComputeResult for entity: Test.7 on thread 139882938156608
UpdateData for entity: Test.2 on thread 139882954942016
UpdateData for entity: Test.3 on thread 139882954942016
ComputeResult for entity: Test.2 on thread 139882954942016
ComputeResult for entity: Test.3 on thread 139882954942016
UpdateData for entity: Test.4 on thread 139882946549312
UpdateData for entity: Test.5 on thread 139882946549312
ComputeResult for entity: Test.4 on thread 139882946549312
ComputeResult for entity: Test.5 on thread 139882946549312
@ Iteration: 2
UpdateData for entity: Test.0 on thread 139882963007296
UpdateData for entity: Test.1 on thread 139882963007296
ComputeResult for entity: Test.0 on thread 139882963007296
ComputeResult for entity: Test.1 on thread 139882963007296
UpdateData for entity: Test.4 on thread 139882946549312
UpdateData for entity: Test.5 on thread 139882946549312
ComputeResult for entity: Test.4 on thread 139882946549312
ComputeResult for entity: Test.5 on thread 139882946549312
UpdateData for entity: Test.2 on thread 139882954942016
UpdateData for entity: Test.3 on thread 139882954942016
ComputeResult for entity: Test.2 on thread 139882954942016
ComputeResult for entity: Test.3 on thread 139882954942016
UpdateData for entity: Test.6 on thread 139882938156608
UpdateData for entity: Test.7 on thread 139882938156608
ComputeResult for entity: Test.6 on thread 139882938156608
ComputeResult for entity: Test.7 on thread 139882938156608
=============== Thread Test Exit ============

[threads_test.txt](https://github.com/SanderMertens/flecs/files/14529238/threads_test.txt)

**Reply by robo-todd** [reply]
The source-example is  a txt so it would let me upload it.

**Reply by robo-todd** [reply]
Also strangely synchronized for each thread as if it is not really parallel, but running different threads at different times. std::cerr for the most visibility I could easily get in outputs. Just really surprised to see pipeline stages running at different times. I guess what I had expected would be to see a bunch of interleaved UpdateData invocations on multiple threads( entity parallel) and then a bunch of interleaved ComputeResult invocations on multiple threads (entity parallel).

**Reply by robo-todd** [reply]
One more bit of information: Essentially the same behavior on v3.2.11.

**Reply by robo-todd** [reply]
This is just a ping. @SanderMertens to see if you have any thoughts?

**Reply by SanderMertens** [reply]
Hey! Sorry for the late reply, I've been busy with an upcoming release. What I think is happening is that the threads just don't have enough to do to fully utilize parallelism- there is quite a bit of context switching overhead which means that for a handful of systems/entities it may not be worth it.

There are known improvements that can reduce the scheduler overhead, but those might take a while to land.

**Reply by robo-todd** [reply]
Thanks for getting back to me. 

So that is not all I'm worried about. In my actual setups I have been assuming all the entities in a given pipeline phase are processed before entities are processed in the next pipeline phase. It is very hard to make a minimal repro example for a full system. The above traces have entities overlapping in phases and I don't quite know how to reason about that from a design perspective.  I can't match up the documentation with the actual behavior so I feel like I don't understand how this works or how to use it.

In my actual application, systems access multiple input entities (pull in data from a query and some traversal of relationships) and update output components assuming that entities are updated in pipeline order. If they are not updated in pipeline order then I don't know how to use the scheduler to design something effective.

**Reply by SanderMertens** [reply]
Phases are not synchronization boundaries. The scheduler checks for system component access + read/write annotations for components that aren't matched by the system query, and uses that to insert sync points. (see this part of the docs for more information: https://www.flecs.dev/flecs/md_docs_2Systems.html#sync-points and this example: https://github.com/SanderMertens/flecs/blob/master/examples/cpp/systems/sync_point/src/main.cpp)

If the scheduler detects no conditions that require a mid-frame sync, it will only happen at the end of the frame. You can see when syncs happen if you connect the explorer to your application, and open the pipeline view.

**Comment by obfuscate** [comment]
Hello there! Does flecs have a way to setup early created threads instead of just setup the number of threads?
